tests/expr/str/test-slice.py

```diff
@@ -0,0 +1,174 @@
+# copy of this software and associated documentation files (the "Software"),
+# to deal in the Software without restriction, including without limitation
+# the rights to use, copy, modify, merge, publish, distribute, sublicense,
+# and/or sell copies of the Software, and to permit persons to whom the
+# Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+# IN THE SOFTWARE.
+#-------------------------------------------------------------------------------
+import pytest
+import random
+from datatable import dt, f, FExpr
+from tests import assert_equals
+
+
+def test_repr():
+    assert str(f.a[:]) == "FExpr<(f.a)[:]>"
+    assert str(f.b[3:]) == "FExpr<(f.b)[3:]>"
+    assert str(f.c[:4]) == "FExpr<(f.c)[:4]>"
+    assert str(f.d[1:4]) == "FExpr<(f.d)[1:4]>"
+    assert str(f.e[::-1]) == "FExpr<(f.e)[::-1]>"
+    assert str(f.f[-1::-1]) == "FExpr<(f.f)[-1::-1]>"
+    assert str(f.g[:4:5]) == "FExpr<(f.g)[:4:5]>"
+    assert str(f.h[3:4:5]) == "FExpr<(f.h)[3:4:5]>"
+    assert str(f.A[0:1]) == "FExpr<(f.A)[0:1]>"
+    assert str((f.A + f.B)[:32]) == "FExpr<(f.A + f.B)[:32]>"
+    assert str(f.S[:-3] + f.S[-2:]) == "FExpr<(f.S)[:-3] + (f.S)[-2:]>"
+    assert str(f.N[f.a:f.b]) == "FExpr<(f.N)[f.a : f.b]>"
+    assert str(f.N[f.a:-1]) == "FExpr<(f.N)[f.a : -1]>"
+    assert str(f.N[f.a::4]) == "FExpr<(f.N)[f.a :: 4]>"
+    assert str(f.N[::f.s]) == "FExpr<(f.N)[:: f.s]>"
+    assert str(f.N[:f.x:f.s]) == "FExpr<(f.N)[: f.x : f.s]>"
+    assert str(f.N[f.w:f.x:f.s]) == "FExpr<(f.N)[f.w : f.x : f.s]>"
+
+
+def test_slice_function():
+    from datatable.str import slice
+    assert str(slice(f[0], 5, 6)) == "FExpr<(f[0])[5:6]>"
+    assert str(slice(f[3], 1, 2, 3)) == "FExpr<(f[3])[1:2:3]>"
+    assert str(slice(f.X, start=f.y, stop=f.z)) == "FExpr<(f.X)[f.y : f.z]>"
+    assert str(slice(f.X, f.a, None, step=f.b)) == "FExpr<(f.X)[f.a :: f.b]>"
+
+
+@pytest.mark.parametrize('ttype',
+    [dt.bool8, dt.int8, dt.int16, dt.int32, dt.int64, dt.float32, dt.float64,
+     dt.Type.date32, dt.Type.time64, dt.obj64])
+def test_slice_wrong_types(ttype):
+    msg = "Slice expression can only be applied to a column of string type"
+    with pytest.raises(TypeError, match=msg):
+        DT = dt.Frame(A=[None], type=ttype)
+        DT[:, f.A[:1]]
+
+
+@pytest.mark.parametrize('ttype',
+    [dt.bool8, dt.float32, dt.float64, dt.str32, dt.str64,
```

Comment:
Is there any reason not to allow `bool8`? This is not a super useful case, of course.

Reply:
I don't see how this would ever be useful. In general, we have bools, integers, and floats, which are all considered numeric, but otherwise in separate categories.

Reply:
For instance, we have some data that could be both "something" and "@something". And there is a boolean `start` that tells if `s[0] == "@"`. Then we strip the actual data by slicing from the boolean `start`.

Reply:
That would have been a rather confusing expression: `f.s[f.s[0] == "@" :]`. But you can always force a boolean expression into integer with `+`: `f.s[+(f.s[0] == "@") :]`, though, admittedly, this doesn't aid readability.

Reply:
The first one doesn't look very unreadable to me. Anyways, if we don't support booleans here - that's not a big deal. If we support, this may be useful in some cases when one wanna make a shortcut.